### PR TITLE
feat: Add multi-turn SFT support

### PR DESCRIPTION
**Pull request by xingyaoww**
# Multi-turn Conversation Fine-tuning Support

## Overview
This PR adds support for fine-tuning models on multi-turn conversations, including proper chat template handling and loss masking for assistant responses. The implementation includes support for the OpenHands SFT dataset and handles conversations up to 32k tokens.

## Key Features
- New `MultiTurnSFTDataset` class for handling multi-turn conversations
- Proper chat template integration using HuggingFace's `apply_chat_template`
- Smart loss masking that targets only assistant responses
- Support for both single-turn and multi-turn training in the same trainer
- Token length limiting and conversation filtering

## Implementation Details
1. Dataset:
   - Uses proper chat templates from the model's tokenizer
   - Handles system, user, and assistant messages
   - Supports conversations with multiple turns
   - Filters conversations exceeding token limits

2. Training:
   - Added `use_multiturn` flag in config
   - Added `messages_key` for multi-turn data format
   - Maintains backward compatibility with single-turn training
   - Works with existing features (FSDP, sequence parallel, etc.)

3. Examples and Tests:
   - Added OpenHands SFT dataset preprocessing script
   - Added multi-turn training example
   - Added comprehensive unit tests
   - Moved tests to appropriate locations

## Usage Example
```yaml
# Config for multi-turn training
data:
  use_multiturn: true
  messages_key: messages
  max_length: 32000
  truncation: right
```

##Testing
- Unit tests for dataset functionality
- Integration with existing training pipeline
- Example scripts tested with OpenHands dataset
- Coverage for both single-turn and multi-turn modes

# Documentation
- Added comments explaining multi-turn specific features
- Updated config defaults with multi-turn options
- Added example scripts with documentation
- Added preprocessing script with dataset-specific handling

---

ok.. this is another PR mostly done by OpenHands with me messaging it about 10 times..

It is still WIP as I'm testing it on a training job now, will report back if it works

### CODE REVIEW

**Comment by xingyaoww** [comment]
Actually i think this PR is ready. I've been using this PR for a while and trained couple model without issue. Would love review here!

Also, do we really need to sign CLA for openhands-agent? 🤣

**Comment by eric-haibin-lin** [comment]
Oh sorry I missed this PR. Will take a look

File: examples/data_preprocess/multiturn.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+    if args.hdfs_dir is not None:
+        makedirs(args.hdfs_dir)
+        copy(src=local_dir, dst=args.hdfs_dir)
+
+    # Print statistics
+    print(f"Train dataset size: {len(train_dataset)}")
+    print(f"Test dataset size: {len(test_dataset)}")
+    print(f"Data saved to {local_dir}")
+EOF > examples / data_preprocess / multiturn.py
```
Comment (eric-haibin-lin) [inline examples/data_preprocess/multiturn.py]: @xingyaoww I wonder why the file is repeated after EOF?

File: verl/utils/dataset/multiturn_sft_dataset.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+        return {
+            'input_ids': input_ids,
+            'attention_mask': attention_mask,
+            'position_ids': position_ids,
+            'loss_mask': loss_mask
+        }
+
+
+EOF > verl / utils / dataset / multiturn_sft_dataset.py
+# Copyright 2024 Bytedance Ltd. and/or its affiliates
```
Comment (eric-haibin-lin) [inline verl/utils/dataset/multiturn_sft_dataset.py]: Same here, repeated after EOF

**Comment by xingyaoww** [comment]
Sorry! Must be some copy-pasta :(

@OpenHands can you help me first merge from main - then help me address these review comments?

File: verl/utils/dataset/multiturn_sft_dataset.py (hunk 2 above)
Comment (xingyaoww) [reply verl/utils/dataset/multiturn_sft_dataset.py]: fixed

File: examples/data_preprocess/multiturn.py (hunk 1 above)
Comment (xingyaoww) [reply examples/data_preprocess/multiturn.py]: fixed

File: verl/trainer/config/sft_trainer.yaml [hunk 3]
```diff
@@ -4,8 +4,12 @@ data:
   micro_batch_size_per_gpu: 4  # this is also val batch size
   train_files: ~/data/gsm8k/train.parquet
   val_files: ~/data/gsm8k/test.parquet
+  # Single-turn settings
   prompt_key: question
   response_key: answer
+  # Multi-turn settings
+  use_multiturn: false  # Set to true to use multi-turn dataset
```
Comment (eric-haibin-lin) [inline verl/trainer/config/sft_trainer.yaml]: what about structure the config this way:
```
multiturn:
    enable: false
    messages_key: message
```
?

**Review by eric-haibin-lin — COMMENTED** [review]
one last comment. otherwise looks good to me!

File: verl/trainer/config/sft_trainer.yaml (hunk 3 above)
Comment (xingyaoww) [reply verl/trainer/config/sft_trainer.yaml]: Done!

**Comment by eric-haibin-lin** [comment]
Is there any benchmark result you want to share on specific datasets?

**Comment by eric-haibin-lin** [comment]
Could u merge main for a fix of megatron tests. And also fix lint with format.sh?

**Review by eric-haibin-lin — APPROVED** [review]
